### PR TITLE
introduce the JSON Processor

### LINKED ISSUE #20052: Add json processor
Add a processor that replaces a field containing json as a string with a field that holds that json as valid json object.

### PR DESCRIPTION
**Pull request by talevy**
Closes #20052

### CODE REVIEW

File: modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/JsonProcessorTests.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+    @SuppressWarnings("unchecked")
+    public void testExecute() throws Exception {
+        String processorTag = randomAsciiOfLength(3);
+        String randomField = randomAsciiOfLength(3);
+        String randomTargetField = randomAsciiOfLength(2);
+        JsonProcessor jsonProcessor = new JsonProcessor(processorTag, randomField, randomTargetField);
+        Map<String, Object> document = new HashMap<>();
+
+        Map<String, Object> randomJsonMap = RandomDocumentPicks.randomSource(random());
```
Comment (talevy) [inline modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/JsonProcessorTests.java]: Something with the following few lines of random json generation is broken. Some XContent work is causing `String`s to become `byte[]` when re-parsed into a Map.

This results in flaky failing tests

Comment (talevy) [inline modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/JsonProcessorTests.java]: NEVERMIND. latest commit fixes this. RandomDocumentPicks was injecting a byte[] into the document... this is not a valid json value and was being encoded as a String by XContent.

I am not sure why we even insert byte arrays

Comment (talevy) [inline modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/JsonProcessorTests.java]: update: it was brought to my attention that CBOR support introduces the ability to have byte arrays in fields: https://github.com/elastic/elasticsearch/commit/da19ddf3e6922a6659cd21c7dcf5deb1dc9bf366. 

I will think of a better way to test this using RandomDocumentPicks without removing the ability to fields to be byte arrays

File: docs/reference/ingest/ingest-node.asciidoc [hunk 2]
```diff
@@ -1270,6 +1270,28 @@ Throws an error when the field is not an array.
...
+| Name           | Required  | Default  | Description
+| `field`        | yes       | -        | The field to be parsed
+| `target_field` | no        | `field`  | The field to insert the converted structured object into
+|======
+
+[source,js]
+--------------------------------------------------
+{
+  "json": {
+    "field": "my_json_string"
```
Comment (tlrx) [inline docs/reference/ingest/ingest-node.asciidoc]: Could it be a real JSON string in doc?

**Comment by talevy** [comment]
@martijnvg mind reviewing this one as well?

File: modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/JsonProcessorTests.java (hunk 1 above)
Comment (talevy) [inline modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/JsonProcessorTests.java]: changed how the test works, this is not an issue anymore

File: test/framework/src/main/java/org/elasticsearch/ingest/IngestDocumentMatcher.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import java.util.Map;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertThat;
+
+/**
+ */
+public class IngestDocumentMatcher {
+    public static void assertIngestDocument(Object a, Object b) {
```
Comment (martijnvg) [inline test/framework/src/main/java/org/elasticsearch/ingest/IngestDocumentMatcher.java]: maybe remove the empty jdocs at the class level and add actual jdoc on the method itself?

**Comment by martijnvg** [comment]
LGTM, assuming build is happy. Left one small comment.

**Comment by talevy** [comment]
thanks! I will wait for the other PR to merge in so that I can remove this `Matcher` here and watch the tests pass

**Pull request merged**